### PR TITLE
Speedy: building transactions with rollback

### PR DESCRIPTION
**Pull request by nickchapman-da**
Speedy: building transactions with rollback

- call `beginTry`, `endTry`, `rollbackTry` and `abortTry` during speedy _on-ledger_ execution
- test: by checking the expected contracts are created by the transaction

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by remyhaemmerle-da — COMMENTED** [review]
Nice !

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -1397,8 +1397,31 @@ private[lf] object SBuiltin {
         case SOptional(opt) =>
           opt match {
             case None =>
+              machine.ledgerMode match {
+                case OffLedger => ()
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: This seems off, we don’t expect to ever hit this do we? I think this should extend `OnLedgerBuiltin` and crash in off-ledger mode.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 2]
```diff
@@ -378,6 +378,11 @@ object SExpr {
     def execute(machine: Machine): Unit = {
       machine.pushKont(KTryCatchHandler(machine, handler))
       machine.ctrl = body
+      machine.ledgerMode match {
+        case OffLedger => ()
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:382]: same here, I think this should extend `OnLegderBuiltin` and crash here.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+  private def typeAndCompile(pkg: Package): PureCompiledPackages = {
+    val rawPkgs = Map(defaultParserParameters.defaultPackageId -> pkg)
+    Validation.checkPackage(rawPkgs, defaultParserParameters.defaultPackageId, pkg)
+    data.assertRight(
+      PureCompiledPackages(rawPkgs, Compiler.Config.Default.copy(stacktracing = FullStackTrace))
+    )
+  }
+
+  private def runUpdateExpr_GetTx(pkgs1: PureCompiledPackages)(e: Expr): SubmittedTransaction = {
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:34]:
```suggestion
  private def runUpdateExprGetTx(pkgs1: PureCompiledPackages)(e: Expr): SubmittedTransaction = {
```
unless there is a good reason to deviate from Scala naming conventions, let’s stick to them.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        template (record : T1) = {
+          precondition True,
+          signatories Cons @Party [(M:T1 {party} record)] (Nil @Party),
+          observers Nil @Party,
+          agreement "Agreement",
+          choices {
+          }
+        };
+
+        val create0 : Party -> Scenario Unit = \(party: Party) ->
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: is there actually a reason why any of these have to go via scenarios? Given that we’d like to get rid of them evenutally, I’d rather avoid building more tests on top of them. You should be able to just call update expressions directly and afterwards call `ptx.finish`

**Review by cocreature — COMMENTED** [review]
lgtm with a few minor comments. Thanks!

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: In general,  I prefer our first level of test depend on the compiler. when test depend on the compiler: 
1 - It make debugging much more involve 
2 - You do not control as well the LF produces, it is maybe not an issue for you compiler guys that have a clear idea how this are compile, but it is not for me. 
3 - A test that tackles a particular feature will survive even if the compiler does not handle it anymore. 

To sum up, in my point of view we should have unit test that do not depends on the compiler **and** integration test.

For sure, we should not rewrite this test now, since it rely on invariant over the cid seeding that will very problably go away.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: I guess not. It was just convenient. I'll try and change to avoid Scenarios.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: Yes, I think you are right.
For some reason I was thinking the exception handling could happen on or off ledger.
But I support the fact that `try`..`catch` is typed as an `Update` prevent this.
I'll change to crash as you suggest.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: I aggree with @moritzkiefer-da. 
However I will not rewrite them now, as the invariants (on cid seeding) they are based will probably go away soon. 
Let's postpone that until we produce actual rollback node.

Comment (cocreature) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: We don’t rely on contract ids in the test do we? We rely on the integer value in the contract which should be independent of any changes here.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala:65]: @cocreature. You are right.

Comment (cocreature) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala]: thanks for making the change!

Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala]: I didn't see this conversation, and already changed to avoid Scenarios.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala]: yup

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: It will happen off ledger but it won’t happen via the speedy builtin.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: What. You can have _off_ ledger execution of `Update` expressions?
I'm confused now!

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: No you can’t have off-ledger execution of `Update`. But you can throw an exception outside of `Update` and we will add a `Script`-specific catch (which won’t go via a speedy builtin, see https://github.com/digital-asset/daml/pull/7368 for an ancient prototype).

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1401]: Ok. that's clear.  A throw can happen anywhere. But the `try-catch` (`Update` expression) can only be on ledger.

And it is the speedy execution of the `try-catch` which performs the `{begin,end,abort,rollback}Try`, which is where the `OnLedger` is required, and so may be assumed.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: done!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 5]
```diff
@@ -1395,11 +1395,28 @@ private[lf] object SBuiltin {
       checkToken(token)
       args.get(0) match {
         case SOptional(opt) =>
-          opt match {
-            case None =>
-              unwindToHandler(machine, payload) //re-throw
-            case Some(handler) =>
-              machine.enterApplication(handler, Array(SEValue(token)))
+          machine.withOnLedger("SBTryHandler") { onLedger =>
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1398]: is there a reason why we cannot extend `OnLedgerBuiltin`?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 6]
```diff
@@ -378,6 +378,9 @@ object SExpr {
     def execute(machine: Machine): Unit = {
       machine.pushKont(KTryCatchHandler(machine, handler))
       machine.ctrl = body
+      machine.withOnLedger("SETryCatch") { onLedger =>
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:381]: same here

**Review by cocreature — APPROVED** [review]
Nice! One minor question left

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 5 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1398]: Yes, probably could here.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala (hunk 6 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:381]: But not here (without some reorganisation) because this is an expression, not a builtin.
